nixos/tests/nextcloud/default.nix

```diff
@@ -1,21 +1,27 @@
 { system ? builtins.currentSystem
 , config ? { }
 , pkgs ? import ../../.. { inherit system config; }
+, lib ? pkgs.lib
 }:
 
-with pkgs.lib;
-
-foldl
+let
+  nextcloudVersions = [ 21 22 23 ];
+  mariadbVersions = [ 104 105 106 ];
+in
+lib.foldl
   (matrix: ver: matrix // {
     "basic${toString ver}" = import ./basic.nix { inherit system pkgs; nextcloudVersion = ver; };
     "with-postgresql-and-redis${toString ver}" = import ./with-postgresql-and-redis.nix {
       inherit system pkgs;
       nextcloudVersion = ver;
     };
-    "with-mysql-and-memcached${toString ver}" = import ./with-mysql-and-memcached.nix {
+  }) { }
+  nextcloudVersions
+// lib.foldl
+  (matrix: ver: matrix // {
+    "with-mysql-and-memcached-nc${toString ver.nextcloudVersion}-mariadb${toString ver.mariadbVersion}" = import ./with-mysql-and-memcached.nix {
       inherit system pkgs;
-      nextcloudVersion = ver;
+      inherit (ver) nextcloudVersion mariadbVersion;
     };
-  })
-{ }
-  [ 21 22 23 ]
+  }) { }
+(lib.cartesianProductOfSets { nextcloudVersion = nextcloudVersions; mariadbVersion = mariadbVersions; })
```

Comment:
I'm not sure if we're overcomplicating things here. Yes, it was a mistake from - well - my side to miss the MariaDB incompatibilities with certain Nextcloud versions, but in the end upstream's job to recommend compatible versions and we should follow that as much as possible. So what I'm suggesting is:

* Revert the change here
* Declare a working mariadb package depending on the Nextcloud version (i.e. Nextcloud 23 and later support 10.6[1], older majors use 10.5) and we only check if this combination from services.nextcloud works fine here.

[1] This is just what I think I recall, I haven't checked that, so no guarantee that this example is 100% correct :)

Reply:
Ok, so how do we suggest we proceed? I can just rebase out the nextcloud change and clean the history up while I'm at it. I'm not overly familiar with nextcloud module and which version of nextcloud is compatble with which mariadb version, so can you maybe do that change?
On the other hand, I'm not sure if we want to default the version of mariadb used on the whole system to something, based on the nextcloud version that's configured. mariadb also doesn't support clean downgrades AFAIR. upgrades also sort-of require manual intervention (running `mysql_upgrade`) rn. not sure if that's documented anywhere.

Reply:
How about setting it with `mkDefault` and perhaps even throw an eval warning if the default is used? (i.e. `mkDefault (lib.warn "using mariadb x.y for nextcloud" pkgs.mariadb_x_y)`?

Reply:
Hm. What would we `mkDefault` to though? 10.5 for all relases < 23? I'm not sure I can commit to maintaining that mapping, tbh. I've rebased the nextcloud changes out, for now.